File: app/src/main/java/com/example/desafioandroid/repositories/PullRepository.java
package com.example.desafioandroid.repositories;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.desafioandroid.feature.endpoints.GetDataService;
import com.example.desafioandroid.feature.utilities.RetrofitClientInstance;
import com.example.desafioandroid.model.Base;
import com.example.desafioandroid.model.Pull;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class PullRepository {

    //Retrofit Interface
    private GetDataService service;

    //Mutable LiveData
    private MutableLiveData<List<Pull>> items;

    public LiveData<List<Pull>> getPulls(String user, String repository){
        if (items == null){
            items = new MutableLiveData<>();
            pulls(user, repository);
        }

        return items;
    }

    private void pulls(String user, String repository){
        service = RetrofitClientInstance.getRetrofitInstance().create(GetDataService.class);
        Call<List<Pull>> call = service.getPullsResponse(user, repository);

        call.enqueue(new Callback<List<Pull>>() {
            @Override
            public void onResponse(Call<List<Pull>> call, Response<List<Pull>> response) {
                List<Pull> pulls = response.body();
                items.setValue(pulls);
            }

            @Override
            public void onFailure(Call<List<Pull>> call, Throwable t) {

            }
        });
    }
}

File: app/src/main/java/com/rafaelpereiraramos/desafioAndroid/repository/PullRepository.kt
package com.rafaelpereiraramos.desafioAndroid.repository

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import androidx.paging.PagedList
import com.rafaelpereiraramos.desafioAndroid.App
import com.rafaelpereiraramos.desafioAndroid.api.GithubService
import com.rafaelpereiraramos.desafioAndroid.database.AppDatabase
import com.rafaelpereiraramos.desafioAndroid.database.dao.PullDAO
import com.rafaelpereiraramos.desafioAndroid.database.model.PullTO
import com.rafaelpereiraramos.desafioAndroid.database.model.RepoTO
import com.rafaelpereiraramos.desafioAndroid.di.Named
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.concurrent.Executor
import javax.inject.Inject

/**
 * Created by Rafael P. Ramos on 14/10/2018.
 */
class PullRepository @Inject constructor(
        val database: AppDatabase,
        val pullDAO: PullDAO,
        @Named("diskIOExecutor") val ioExecutor: Executor,
        @Named("networkExecutor") val networkExecutor: Executor,
        val service: GithubService
) : PagedList.BoundaryCallback<PullTO>() {

    private var repoName: String? = null
    private var ownerLogin: String? = null
    private var lastRequestedPage = 1
    val networkErrors = MutableLiveData<String>()
    private var isRequestInProgress = false

    override fun onZeroItemsLoaded() {
        requestAndSaveData()
    }

    override fun onItemAtEndLoaded(itemAtEnd: PullTO) {
        requestAndSaveData()
    }

    fun insert(pulls: List<PullTO>, insertFinished: ()-> Unit) {
        ioExecutor.execute {
            pulls.forEach {it.repoOwnerLogin = ownerLogin; it.repoName = repoName}
            pullDAO.insert(pulls)
            insertFinished()
        }
    }

    fun getPulls(repo: RepoTO): DataSource.Factory<Int, PullTO> {
        return this.getPulls(repo.owner.login, repo.name)
    }

    fun getPulls(ownerLogin: String, repoName: String): DataSource.Factory<Int, PullTO> {
        this.ownerLogin = ownerLogin
        this.repoName = repoName

        return pullDAO.getPulls(ownerLogin, repoName)
    }

    private fun requestAndSaveData() {
        if (isRequestInProgress) return

        isRequestInProgress = true

        networkExecutor.execute {
            service.getPull(ownerLogin!!, repoName!!, lastRequestedPage, App.NETWORK_SNAPSHOT_SIZE)
                    .enqueue(object : Callback<List<PullTO>> {
                        override fun onFailure(call: Call<List<PullTO>>, t: Throwable) {
                            networkErrors.postValue(t.message)
                            t.printStackTrace()
                            isRequestInProgress = false
                        }

                        override fun onResponse(call: Call<List<PullTO>>, response: Response<List<PullTO>>) {
                            if (response.isSuccessful) {
                                val pulls = response.body() ?: emptyList()
                                insert(pulls) {
                                    lastRequestedPage++
                                    isRequestInProgress = false
                                }
                            } else {
                                networkErrors.postValue(response.errorBody()?.string())
                                isRequestInProgress = false
                            }
                        }
                    })
        }
    }
}